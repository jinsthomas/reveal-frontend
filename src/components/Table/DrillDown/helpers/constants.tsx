/** Constants */
import React from 'react';

export const ID = 'id';
export type ID = typeof ID;

export const PARENT_ID = 'parent_id';
export type PARENT_ID = typeof PARENT_ID;

export const ROOT_PARENT_ID = null;
export type ROOT_PARENT_ID = typeof ROOT_PARENT_ID;

export const LINKER_ITEM_CSS_CLASS = 'dd-linker-item';
export type LINKER_ITEM_CSS_CLASS = typeof LINKER_ITEM_CSS_CLASS;

export const CLICKABLE_CSS_CLASS = 'dd-linker-item dd-clickable';
export type CLICKABLE_CSS_CLASS = typeof CLICKABLE_CSS_CLASS;

export const CARET_CSS_CLASS = 'dd-caret';
export type CARET_CSS_CLASS = typeof CARET_CSS_CLASS;

export const CARET_SPAN = <span className={CARET_CSS_CLASS}>&nbsp;&#9660;</span>;
export type CARET_SPAN = typeof CARET_SPAN;

export const DEFAULT_ROW_HEIGHT = '2em';
export type DEFAULT_ROW_HEIGHT = typeof DEFAULT_ROW_HEIGHT;
