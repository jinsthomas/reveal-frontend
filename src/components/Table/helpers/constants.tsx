/** Constants */
import React from 'react';
export const ID = 'id';
export type ID = typeof ID;

export const PARENT_ID = 'parent_id';
export type PARENT_ID = typeof PARENT_ID;

export const ROOT_PARENT_ID = null;
export type ROOT_PARENT_ID = typeof ROOT_PARENT_ID;

export const LINKER_ITEM_CSS_CLASS = 'dd-linker-item';
export type LINKER_ITEM_CSS_CLASS = typeof LINKER_ITEM_CSS_CLASS;

export const CLICKABLE_CSS_CLASS = 'dd-linker-item dd-clickable';
export type CLICKABLE_CSS_CLASS = typeof CLICKABLE_CSS_CLASS;

export const CARET_CSS_CLASS = 'dd-caret';
export type CARET_CSS_CLASS = typeof CARET_CSS_CLASS;

export const CARET = <span className={CARET_CSS_CLASS}>&nbsp;&#9660;</span>;
export type CARET = typeof CARET;

export const ROWS_TO_DISPLAY = 'Rows to display';
export type ROWS_TO_DISPLAY = typeof ROWS_TO_DISPLAY;

export const PAGE = 'page';
export type PAGE = typeof PAGE;

export const OF = 'of';
export type OF = typeof OF;
