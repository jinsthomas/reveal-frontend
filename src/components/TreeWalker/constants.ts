export const ACTIVE = 'Active';
export const FIND_BY_ID = 'location/findByJurisdictionIds';
export const FIND_BY_PROPERTIES = 'location/findByProperties';
export const LOCATION = 'location';
export const COULDNT_LOAD_PARENTS = 'Could not load parents';
