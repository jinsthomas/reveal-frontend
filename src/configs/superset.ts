export const SUPERSET_API_BASE = 'https://discover.ona.io/';
export type SUPERSET_API_BASE = typeof SUPERSET_API_BASE;

export const SUPERSET_API_ENDPOINT = 'slice';
export type SUPERSET_API_ENDPOINT = typeof SUPERSET_API_ENDPOINT;
