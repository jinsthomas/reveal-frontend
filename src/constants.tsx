// env vars
export const WEBSITE_NAME = process.env.REACT_APP_WEBSITE_NAME;
export type WEBSITE_NAME = typeof WEBSITE_NAME;

// strings
export const LOCATION = 'Location';
export type LOCATION = typeof LOCATION;
export const PROVINCE = 'Province';
export type PROVINCE = typeof PROVINCE;
export const HOME = 'Home';
export type HOME = typeof HOME;
export const FOCUS_INVESTIGATION = 'Focus Investigation';
export type FOCUS_INVESTIGATION = typeof FOCUS_INVESTIGATION;
export const FOCUS_INVESTIGATIONS = 'Focus Investigations';
export type FOCUS_INVESTIGATIONS = typeof FOCUS_INVESTIGATIONS;

// internal urls
export const HOME_URL = '/';
export type HOME_URL = typeof HOME_URL;
export const IRS_URL = '/irs';
export type IRS_URL = typeof IRS_URL;
export const FI_URL = '/focus-investigation';
export type FI_URL = typeof FI_URL;
export const FI_HISTORICAL_URL = '/focus-investigation/historical';
export type FI_HISTORICAL_URL = typeof FI_HISTORICAL_URL;

// colors
export const GREEN = 'Green';
export type GREEN = typeof GREEN;
export const YELLOW = 'Yellow';
export type YELLOW = typeof YELLOW;
export const ORANGE = 'Orange';
export type ORANGE = typeof ORANGE;
export const RED = 'Red';
export type RED = typeof RED;

// thresholds
export const GREEN_THRESHOLD = 0.9;
export type GREEN_THRESHOLD = typeof GREEN_THRESHOLD;
export const YELLOW_THRESHOLD = 0.2;
export type YELLOW_THRESHOLD = typeof YELLOW_THRESHOLD;
export const ORANGE_THRESHOLD = 0.8;
export type ORANGE_THRESHOLD = typeof ORANGE_THRESHOLD;

// 1-3-7 thresholds
export const THRESHOLD_137 = 0;
