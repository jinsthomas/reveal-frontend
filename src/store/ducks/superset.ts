import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** the reducer name */
export const reducerName = 'superset';

/** interface for authorize action */
interface AuthorizeSupersetAction extends AnyAction {
  authorized: boolean;
  type: typeof AUTHORIZE;
}

/** Create type for Superset reducer actions */
export type SupersetActionTypes = AuthorizeSupersetAction | AnyAction;

/** interface for superset state */
interface SupersetState {
  authorized: boolean | null;
}

/** immutable session state */
export type ImmutableSupersetState = SupersetState &
  SeamlessImmutable.ImmutableObject<SupersetState>;

/** initial state */
const initialState: ImmutableSupersetState = SeamlessImmutable({
  authorized: null,
});

/** the superset reducer function */
export default function reducer(state = initialState, action: SupersetActionTypes): SupersetState {
  switch (action.type) {
    case AUTHORIZE:
      return state.merge({
        authorized: action.authorized,
      });
    default:
      return state;
  }
}

// actions
/** authorize action type */
export const AUTHORIZE = 'reveal/reducer/superset/AUTHORIZE';

// action creators

/** authorize action creator
 * @param {boolean} authorized - whether superset is authorized or not
 */
export const authorizeSuperset = (authorized: boolean): AuthorizeSupersetAction => ({
  authorized,
  type: AUTHORIZE,
});

// selectors

/** check if superset is authorized
 * @param {Partial<Store>} state - the redux store
 */
export function isAuthorized(state: Partial<Store>): boolean | null {
  return (state as any)[reducerName].authorized;
}
