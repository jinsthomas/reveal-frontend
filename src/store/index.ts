import { combine } from '@onaio/redux-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { applyMiddleware, compose, createStore, Reducer } from 'redux';
import thunk from 'redux-thunk';
import users, { reducerName as usersReducer } from './ducks/users';

/** Add redux devtools to Window */
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

/** Declare history variable */
export const history = createBrowserHistory();

/** Declare type for reducer Registry */
export interface Registry {
  [key: string]: Reducer;
}

/** Function to create the store */
export function getStore(reducers: Registry) {
  /** Register each of the initial reducers */
  Object.keys(reducers).forEach(reducerName => {
    reducerRegistry.register(reducerName, reducers[reducerName]);
  });

  /** Combine reducers */
  const reducer = combine(reducerRegistry.getReducers());

  /** Add redux devtools to enhancers */
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  /** Create the store */
  return createStore(reducer, composeEnhancers(applyMiddleware(thunk, routerMiddleware(history))));
}

/** Router reducer */
const connectReducer = connectRouter(history);

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[usersReducer] = users;

/** The initial store for the reveal web app */
const store = getStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
  store.replaceReducer(combine(reducers));
});

export default store;
