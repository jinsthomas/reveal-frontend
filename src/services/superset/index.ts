import superset from '@onaio/superset-connector';
import { clone } from 'lodash';
import { CONFIG } from '../../configs/superset';
import store from '../../store';
import { getApiToken } from '../../store/selectors';

/** middleware for fetching from Superset */
export const fetchMiddleware = (res: { [key: string]: any }) => {
  return res;
};

/** callback for fetching from Superset */
export const fetchCallback = (parsedResponse: Array<{ [key: string]: any }>) => {
  const sliceData = superset.processData(parsedResponse);
};

/** generic async function for fetching from superset */
const supersetFetch = async (
  sliceId: string,
  callback: typeof fetchCallback = fetchCallback,
  middleware: typeof fetchMiddleware = fetchMiddleware
) => {
  const config = clone(CONFIG);
  config.extraPath = sliceId;
  config.token = getApiToken(store.getState());
  superset.api.fetch(config, middleware).then(callback);
};

export default supersetFetch;
