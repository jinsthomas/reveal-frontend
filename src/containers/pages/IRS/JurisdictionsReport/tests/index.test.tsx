import reducerRegistry from '@onaio/redux-reducer-registry';
import superset from '@onaio/superset-connector';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Helmet } from 'react-helmet';
import { Router } from 'react-router';
import { JurisdictionReport } from '../';
import { INTERVENTION_IRS_URL, IRS_REPORTING_TITLE, MAP } from '../../../../../constants';
import store from '../../../../../store';
import GenericJurisdictionsReducer, {
  fetchGenericJurisdictions,
  GenericJurisdiction,
  getGenericJurisdictionByJurisdictionId,
  getGenericJurisdictionsArray,
  reducerName as GenericJurisdictionsReducerName,
} from '../../../../../store/ducks/generic/jurisdictions';
import IRSPlansReducer, {
  fetchIRSPlans,
  getIRSPlanById,
  IRSPlan,
  reducerName as IRSPlansReducerName,
} from '../../../../../store/ducks/generic/plans';
import genericStructuresReducer, {
  fetchGenericStructures,
  GenericStructure,
  getGenericStructures,
  reducerName as genericStructuresReducerName,
  StructureFeatureCollection,
} from '../../../../../store/ducks/generic/structures';
import { plans } from '../../../../../store/ducks/generic/tests/fixtures';
import jurisdictionReducer, {
  fetchJurisdictions,
  getJurisdictionById,
  Jurisdiction,
  reducerName as jurisdictionReducerName,
} from '../../../../../store/ducks/jurisdictions';
import * as fixtures from '../../JurisdictionsReport/tests/fixtures';

/* tslint:disable-next-line no-var-requires */
const fetch = require('jest-fetch-mock');

jest.mock('../../../../../configs/env');

/** register the reducers */
reducerRegistry.register(IRSPlansReducerName, IRSPlansReducer);
reducerRegistry.register(GenericJurisdictionsReducerName, GenericJurisdictionsReducer);

/** set up data in the store needed for this view */
const focusAreaData = superset.processData(fixtures.ZambiaFocusAreasJSON) || [];
const jurisdictionData = superset.processData(fixtures.ZambiaJurisdictionsJSON) || [];

store.dispatch(fetchGenericJurisdictions('zm-jurisdictions', jurisdictionData));
store.dispatch(fetchGenericJurisdictions('zm-focusAreas', focusAreaData));

const history = createBrowserHistory();

describe('components/IRS Reports/JurisdictionReport', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const mock: any = jest.fn();
    let jurisdictions =
      getGenericJurisdictionsArray(
        store.getState(),
        'zm-jurisdictions',
        '9f1e0cfa-5313-49ff-af2c-f7dbf4fbdb9d'
      ) || [];
    jurisdictions = jurisdictions.concat(
      getGenericJurisdictionsArray(
        store.getState(),
        'zm-focusAreas',
        '9f1e0cfa-5313-49ff-af2c-f7dbf4fbdb9d'
      )
    );
    const props = {
      history,
      jurisdictions,
      location: mock,
      match: {
        isExact: true,
        params: {
          planId: (plans[0] as IRSPlan).plan_id,
        },
        path: `${INTERVENTION_IRS_URL}/:planId`,
        url: `${INTERVENTION_IRS_URL}/${(plans[0] as IRSPlan).plan_id}`,
      },
      plan: plans[0] as IRSPlan,
    };
    shallow(
      <Router history={history}>
        <JurisdictionReport {...props} />
      </Router>
    );
  });

  it('renders correctly', async () => {
    fetch.mockResponseOnce(JSON.stringify({}));
    const mock: any = jest.fn();

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => []);

    let jurisdictions =
      getGenericJurisdictionsArray(
        store.getState(),
        'zm-jurisdictions',
        '9f1e0cfa-5313-49ff-af2c-f7dbf4fbdb9d'
      ) || [];
    jurisdictions = jurisdictions.concat(
      getGenericJurisdictionsArray(
        store.getState(),
        'zm-focusAreas',
        '9f1e0cfa-5313-49ff-af2c-f7dbf4fbdb9d'
      )
    );

    const props = {
      history,
      jurisdictions,
      location: mock,
      match: {
        isExact: true,
        params: {
          planId: (plans[0] as IRSPlan).plan_id,
        },
        path: `${INTERVENTION_IRS_URL}/:planId`,
        url: `${INTERVENTION_IRS_URL}/${(plans[0] as IRSPlan).plan_id}`,
      },
      plan: plans[0] as IRSPlan,
      service: supersetServiceMock,
    };

    const wrapper = mount(
      <Router history={history}>
        <JurisdictionReport {...props} />
      </Router>
    );
    const helmet = Helmet.peek();
    await flushPromises();
    expect(toJson(wrapper.find('BreadcrumbItem li'))).toMatchSnapshot('breadcrumbs');
    expect(toJson(wrapper.find('h3.page-title'))).toMatchSnapshot('page title');
    expect(helmet.title).toEqual('IRS Reporting: IRS 2019-09-05 TEST');
    expect(wrapper.find('DrillDownTable').props()).toMatchSnapshot({
      data: expect.any(Object) /** just for purposes of making snapshot smaller */,
    });
    expect(wrapper.find('DrillDownTable').props().data).toEqual(jurisdictions);
    expect(supersetServiceMock.mock.calls).toEqual([
      [
        '13',
        {
          adhoc_filters: [
            {
              clause: 'WHERE',
              comparator: '727c3d40-e118-564a-b231-aac633e6abce',
              expressionType: 'SIMPLE',
              operator: '==',
              subject: 'plan_id',
            },
          ],
          row_limit: 1,
        },
      ],
      [
        '11',
        {
          adhoc_filters: [
            {
              clause: 'WHERE',
              comparator: '727c3d40-e118-564a-b231-aac633e6abce',
              expressionType: 'SIMPLE',
              operator: '==',
              subject: 'plan_id',
            },
          ],
          order_by_cols: ['["jurisdiction_depth",+true]'],
          row_limit: 3000,
        },
      ],
      [
        '12',
        {
          adhoc_filters: [
            {
              clause: 'WHERE',
              comparator: '727c3d40-e118-564a-b231-aac633e6abce',
              expressionType: 'SIMPLE',
              operator: '==',
              subject: 'plan_id',
            },
          ],
          order_by_cols: ['["jurisdiction_depth",+true]'],
          row_limit: 3000,
        },
      ],
    ]);
    expect(supersetServiceMock).toHaveBeenCalledTimes(3);
    wrapper.unmount();
  });
});
