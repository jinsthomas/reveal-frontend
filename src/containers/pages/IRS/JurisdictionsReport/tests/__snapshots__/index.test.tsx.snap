// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/IRS Reports/JurisdictionReport renders correctly: breadcrumbs 1`] = `
Array [
  <li
    className="breadcrumb-item"
  >
    <Link
      key="0"
      replace={false}
      to="/"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
    </Link>
  </li>,
  <li
    className="breadcrumb-item"
  >
    <Link
      key="1"
      replace={false}
      to="/intervention/irs/report"
    >
      <a
        href="/intervention/irs/report"
        onClick={[Function]}
      >
        IRS Reporting
      </a>
    </Link>
  </li>,
  <li
    aria-current="page"
    className="active breadcrumb-item"
  >
    IRS 2019-09-05 TEST
  </li>,
]
`;

exports[`components/IRS Reports/JurisdictionReport renders correctly: page title 1`] = `
<h3
  className="mb-3 page-title"
>
  IRS Reporting: IRS 2019-09-05 TEST
</h3>
`;
